pkg/query/remote_engine.go

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package query
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/util/stats"
+	"github.com/thanos-community/promql-engine/api"
+
+	"github.com/thanos-io/thanos/pkg/api/query/querypb"
+	"github.com/thanos-io/thanos/pkg/store/labelpb"
+)
+
+// Opts are the options for a PromQL query.
+type Opts struct {
+	AutoDownSample        bool
+	ReplicaLabels         []string
+	Timeout               time.Duration
+	EnablePartialResponse bool
+}
+
+// Client is a query client that executes PromQL queries.
+type Client struct {
+	querypb.QueryClient
+	address   string
+	maxt      int64
+	labelSets []labels.Labels
+}
+
+// NewClient creates a new Client.
+func NewClient(queryClient querypb.QueryClient, address string, maxt int64, labelSets []labels.Labels) Client {
+	return Client{QueryClient: queryClient, address: address, maxt: maxt, labelSets: labelSets}
+}
+
+func (q Client) MaxT() int64 { return q.maxt }
+
+func (q Client) LabelSets() []labels.Labels { return q.labelSets }
+
+func (q Client) GetAddress() string { return q.address }
+
+type remoteEndpoints struct {
+	logger     log.Logger
+	getClients func() []Client
+	opts       Opts
+}
+
+func NewRemoteEndpoints(logger log.Logger, getClients func() []Client, opts Opts) api.RemoteEndpoints {
+	return remoteEndpoints{
+		logger:     logger,
+		getClients: getClients,
+		opts:       opts,
+	}
+}
+
+func (r remoteEndpoints) Engines() []api.RemoteEngine {
+	clients := r.getClients()
+	engines := make([]api.RemoteEngine, len(clients))
+	for i := range clients {
+		engines[i] = newRemoteEngine(r.logger, clients[i], r.opts)
+	}
+	return engines
+}
+
+type remoteEngine struct {
+	Client
+	opts   Opts
+	logger log.Logger
+}
+
+func newRemoteEngine(logger log.Logger, queryClient Client, opts Opts) api.RemoteEngine {
+	return &remoteEngine{
+		logger: logger,
+		Client: queryClient,
+		opts:   opts,
+	}
+}
+
+func (r remoteEngine) LabelSets() []labels.Labels {
+	replicaLabelSet := make(map[string]struct{})
+	for _, lbl := range r.opts.ReplicaLabels {
+		replicaLabelSet[lbl] = struct{}{}
+	}
+
+	// Strip replica labels from the result.
+	labelSets := r.Client.LabelSets()
+	result := make([]labels.Labels, len(labelSets))
+	for i := range labelSets {
+		numLabels := len(labelSets[i]) - len(r.opts.ReplicaLabels)
```

Comment:
I'm not sure if this logic is right. It should strip replica labels, right? What if `labelSets[i]` is:

```
[a=b, c=d]
```

And `replicaLabels` is `[e, f, g]`.

In this case, `result[i]` will be empty even if it should be `[a=b, c=d]`? Maybe it would be better to use `labels.Builder` here to avoid bugs in logic?

Reply:
You are right, this was indeed a bug. I changed the logic to use the newly added `labels.ScratchBuilder` instead and added a test for this.